File: ldk/csharp/OliveHelpsLDK/OliveHelpsLDK/Filesystem/FilesystemClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Grpc.Core;
using OliveHelpsLDK.Logging;
using Proto;

namespace OliveHelpsLDK.Filesystem
{
    internal class FilesystemClient : BaseClient<Proto.Filesystem.FilesystemClient>, IFilesystemService
    {
        internal FilesystemClient(Proto.Filesystem.FilesystemClient client, Session session, ILogger logger) : base(
            client, session, logger, "filesystem")
        {
        }

        internal FilesystemClient(CallInvoker callInvoker, Session session, ILogger logger) : this(
            new Proto.Filesystem.FilesystemClient(callInvoker), session, logger)
        {
        }

        public Task<IList<FileInfo>> QueryDirectory(string directoryPath,
            CancellationToken cancellationToken = default)
        {
            var request = new FilesystemDirRequest
            {
                Session = CreateSession(),
                Directory = directoryPath,
            };
            var loggedParser =
                LoggedParser<Task<FilesystemDirResponse>, IList<FileInfo>>(response =>
                    ConvertFileList(response.Result.Files));
            return Client.FilesystemDirAsync(request, CreateOptions(cancellationToken))
                .ResponseAsync
                .ContinueWith(loggedParser, cancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion,
                    TaskScheduler.Current);
        }

        public IStreamingCall<FileEvent> StreamDirectory(string directoryPath,
            CancellationToken cancellationToken = default)
        {
            var request = new FilesystemDirStreamRequest
            {
                Session = CreateSession(),
                Directory = directoryPath
            };
            var call = Client.FilesystemDirStream(request, CreateOptions(cancellationToken));
            var loggedParser = LoggedParser<FilesystemDirStreamResponse, FileEvent>(ConvertFileEvent);
            return new StreamingCall<FilesystemDirStreamResponse, FileEvent>(call, loggedParser);
        }

        public IStreamingCall<FileEvent> StreamFileInfo(string filePath, CancellationToken cancellationToken = default)
        {
            var request = new FilesystemFileInfoStreamRequest
            {
                Session = CreateSession(),
                Path = filePath
            };
            var call = Client.FilesystemFileInfoStream(request, CreateOptions(cancellationToken));
            var loggedParser = LoggedParser<FilesystemFileInfoStreamResponse, FileEvent>(ConvertFileEvent);
            return new StreamingCall<FilesystemFileInfoStreamResponse, FileEvent>(call, loggedParser);
        }

        public Task MakeDirectory(MakeDirectoryParameters parameters, CancellationToken cancellationToken = default)
        {
            var request = new FilesystemMakeDirRequest
            {
                Session = CreateSession(),
                Path = parameters.Path,
                Perm = parameters.Permissions
            };
            return Client.FilesystemMakeDirAsync(request, CreateOptions(cancellationToken)).ResponseAsync;
        }

        public Task Copy(MoveOrCopyFileParameters parameters, CancellationToken cancellationToken = default)
        {
            var request = new FilesystemCopyRequest()
            {
                Session = CreateSession(),
                Source = parameters.Source,
                Dest = parameters.Destination
            };
            return Client.FilesystemCopyAsync(request, CreateOptions(cancellationToken)).ResponseAsync;
        }

        public Task Move(MoveOrCopyFileParameters parameters, CancellationToken cancellationToken = default)
        {
            var request = new FilesystemMoveRequest()
            {
                Session = CreateSession(),
                Source = parameters.Source,
                Dest = parameters.Destination
            };
            return Client.FilesystemMoveAsync(request, CreateOptions(cancellationToken)).ResponseAsync;
        }

        public Task Remove(RemoveParameters parameters, CancellationToken cancellationToken = default)
        {
            var request = new FilesystemRemoveRequest()
            {
                Session = CreateSession(),
                Path = parameters.Path,
                Recursive = parameters.Recursive
            };
            return Client.FilesystemRemoveAsync(request, CreateOptions(cancellationToken)).ResponseAsync;
        }

        public async Task<IFile> OpenFile(string path, CancellationToken cancellationToken = default)
        {
            var stream = Client.FilesystemFileStream(CreateOptions(cancellationToken));
            var file = new File(stream, Logger, Session);
            await file.Open(path);
            return file;
        }

        public async Task<IFile> CreateFile(string path, CancellationToken cancellationToken = default)
        {
            var stream = Client.FilesystemFileStream(CreateOptions(cancellationToken));
            var file = new File(stream, Logger, Session);
            await file.Create(path);
            return file;
        }

        private static IList<FileInfo> ConvertFileList(IEnumerable<Proto.FileInfo> files)
        {
            return files.Select(ConvertProtoFileInfo).ToList();
        }

        private static FileInfo ConvertProtoFileInfo(Proto.FileInfo info)
        {
            return new FileInfo
            {
                IsDirectory = info.IsDir,
                Name = info.Name,
                Mode = info.Mode,
                Size = info.Size,
                Updated = info.Updated.ToDateTime(),
            };
        }

        private static FileEvent ConvertFileEvent(FilesystemDirStreamResponse dirStreamResponse)
        {
            return new FileEvent
            {
                FileInfo = ConvertProtoFileInfo(dirStreamResponse.File),
                FileAction = ConvertFileAction(dirStreamResponse.Action),
            };
        }

        private static FileEvent ConvertFileEvent(FilesystemFileInfoStreamResponse fileStreamResponse)
        {
            return new FileEvent
            {
                FileInfo = ConvertProtoFileInfo(fileStreamResponse.File),
                FileAction = ConvertFileAction(fileStreamResponse.Action),
            };
        }

        private static FileAction ConvertFileAction(Proto.FileAction action)
        {
            return action switch
            {
                Proto.FileAction.Unknown => FileAction.Unknown,
                Proto.FileAction.Create => FileAction.Create,
                Proto.FileAction.Write => FileAction.Write,
                Proto.FileAction.Remove => FileAction.Remove,
                Proto.FileAction.Rename => FileAction.Rename,
                Proto.FileAction.Chmod => FileAction.Chmod,
                _ => throw new ArgumentOutOfRangeException(nameof(action), action, null)
            };
        }
    }
}

File: ldk/node/src/loopClients/fileSystemClient.ts
import { Empty } from 'google-protobuf/google/protobuf/empty_pb';
import { FilesystemClient as FilesystemGRPCClient } from '../grpc/filesystem_grpc_pb';
import messages, { FileAction } from '../grpc/filesystem_pb';
import BaseClient, { GRPCClientConstructor } from './baseClient';
import {
  FileSystemCopyOrMoveParams,
  FileSystemFile,
  FileSystemMakeDirectoryParams,
  FileSystemQueryDirectoryParams,
  FileSystemQueryDirectoryResponse,
  FileSystemQueryFileParams,
  FileSystemRemoveParams,
  FileSystemSensor,
  FileSystemStreamAction,
  FileSystemStreamDirectoryResponse,
  FileSystemStreamFileInfoResponse,
} from './fileSystemSensor';
import { StoppableStream, StreamListener } from './stoppables';
import { TransformingStream } from './transformingStream';
import { FileSystemFileImpl, parseFileInfo } from './fileSystemFile';

/**
 * @param action - The file action.
 * @internal
 */
function parseFileAction(action: FileAction): FileSystemStreamAction {
  switch (action) {
    case FileAction.FILE_ACTION_CREATE:
      return FileSystemStreamAction.Create;
    case FileAction.FILE_ACTION_WRITE:
      return FileSystemStreamAction.Write;
    case FileAction.FILE_ACTION_REMOVE:
      return FileSystemStreamAction.Remove;
    case FileAction.FILE_ACTION_RENAME:
      return FileSystemStreamAction.Rename;
    case FileAction.FILE_ACTION_CHMOD:
      return FileSystemStreamAction.Chmod;
    case FileAction.FILE_ACTION_UNKNOWN:
    default:
      return FileSystemStreamAction.Unknown;
  }
}

/**
 * @internal
 */
export class FileSystemClient
  extends BaseClient<FilesystemGRPCClient>
  implements FileSystemSensor {
  protected generateClient(): GRPCClientConstructor<FilesystemGRPCClient> {
    return FilesystemGRPCClient;
  }

  queryDirectory(
    params: FileSystemQueryDirectoryParams,
  ): Promise<FileSystemQueryDirectoryResponse> {
    return this.buildQuery<
      messages.FilesystemDirRequest,
      messages.FilesystemDirResponse,
      FileSystemQueryDirectoryResponse
    >(
      (message, callback) => {
        this.client.filesystemDir(message, callback);
      },
      () => new messages.FilesystemDirRequest().setDirectory(params.directory),
      (message) => ({
        files: message.getFilesList().map(parseFileInfo),
      }),
    );
  }

  streamDirectory(
    params: FileSystemQueryDirectoryParams,
    listener: StreamListener<FileSystemStreamDirectoryResponse>,
  ): StoppableStream<FileSystemStreamDirectoryResponse> {
    const message = new messages.FilesystemDirStreamRequest()
      .setDirectory(params.directory)
      .setSession(this.createSessionMessage());
    return new TransformingStream<
      messages.FilesystemDirStreamResponse,
      FileSystemStreamDirectoryResponse
    >(
      this.client.filesystemDirStream(message),
      (response) => {
        const fileInfo = response.getFile();
        if (fileInfo == null) {
          return undefined;
        }
        return {
          files: parseFileInfo(fileInfo),
          action: parseFileAction(response.getAction()),
        };
      },
      listener,
    );
  }

  streamFileInfo(
    params: FileSystemQueryFileParams,
    listener: StreamListener<FileSystemStreamFileInfoResponse>,
  ): StoppableStream<FileSystemStreamFileInfoResponse> {
    const message = new messages.FilesystemFileInfoStreamRequest()
      .setPath(params.file)
      .setSession(this.createSessionMessage());
    return new TransformingStream<
      messages.FilesystemFileInfoStreamResponse,
      FileSystemStreamFileInfoResponse
    >(
      this.client.filesystemFileInfoStream(message),
      (response) => {
        const fileInfo = response.getFile();
        if (fileInfo == null) {
          return undefined;
        }
        return {
          file: parseFileInfo(fileInfo),
          action: parseFileAction(response.getAction()),
        };
      },
      listener,
    );
  }

  copyFile(params: FileSystemCopyOrMoveParams): Promise<void> {
    const message = new messages.FilesystemCopyRequest()
      .setDest(params.destination)
      .setSource(params.source)
      .setSession(this.createSessionMessage());
    return this.buildQuery<messages.FilesystemCopyRequest, Empty, void>(
      (request, callback) => {
        this.client.filesystemCopy(request, callback);
      },
      () => message,
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      () => {},
    );
  }

  moveFile(params: FileSystemCopyOrMoveParams): Promise<void> {
    const message = new messages.FilesystemMoveRequest()
      .setDest(params.destination)
      .setSource(params.source)
      .setSession(this.createSessionMessage());
    return this.buildQuery<messages.FilesystemMoveRequest, Empty, void>(
      (request, callback) => {
        this.client.filesystemMove(request, callback);
      },
      () => message,
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      () => {},
    );
  }

  makeDirectory(path: FileSystemMakeDirectoryParams): Promise<void> {
    return this.buildQuery<messages.FilesystemMakeDirRequest, Empty, void>(
      (request, callback) => {
        this.client.filesystemMakeDir(request, callback);
      },
      () =>
        new messages.FilesystemMakeDirRequest()
          .setPath(path.path)
          .setPerm(path.permissions)
          .setSession(this.createSessionMessage()),
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      () => {},
    );
  }

  openFile(path: string): FileSystemFile {
    const impl = new FileSystemFileImpl(
      this.session,
      this.client.filesystemFileStream(),
      this.logger,
    );
    impl.open(path);
    return impl;
  }

  createFile(path: string): FileSystemFile {
    const impl = new FileSystemFileImpl(
      this.session,
      this.client.filesystemFileStream(),
      this.logger,
    );
    impl.create(path);
    return impl;
  }

  removeFile(params: FileSystemRemoveParams): Promise<void> {
    return this.buildQuery<messages.FilesystemRemoveRequest, Empty, void>(
      (request, callback) => {
        this.client.filesystemRemove(request, callback);
      },
      () =>
        new messages.FilesystemRemoveRequest()
          .setPath(params.path)
          .setRecursive(params.recursive || false)
          .setSession(this.createSessionMessage()),
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      () => {},
    );
  }

  protected sensorName(): string {
    return 'filesystem';
  }
}
